File: examples/typescript/react/notes-showcase/src/viewmodels/notificationsVM.ts
/**
 * NotificationsVM — subscribes to an INotificationHub and surfaces the most
 * recent notifications as a bounded list of NotificationVM (cap = 5).
 *
 * VMx-API adaptation: VMx's `NotificationVM` requires a `SchedulerLike`
 * (RxJS) and a lifespan; passing `asyncScheduler` + a configurable lifespan
 * lets tests inject `VirtualTimeScheduler` for deterministic timer control
 * (per the cross-language note re Python's TimeoutScheduler / TestScheduler).
 */
import { asyncScheduler, type SchedulerLike } from "rxjs";
import {
  ComponentVMBase,
  PropertyChangedMessage,
  ViewModelType,
  type IDispatcher,
  type IMessageHub,
} from "vmx";
import {
  type INotificationHub,
  type Notification,
  NotificationVM,
} from "vmx/notifications";

export class NotificationsVM extends ComponentVMBase {
  static readonly DEFAULT_CAP = 5;

  readonly #notificationHub: INotificationHub;
  readonly #scheduler: SchedulerLike;
  readonly #lifespanMs: number | undefined;
  readonly #cap: number;
  readonly #visible: NotificationVM[] = [];
  readonly #map = new Map<Notification, NotificationVM>();
  #pendingSub: { unsubscribe(): void } | null = null;

  constructor(opts: {
    name: string;
    hint: string;
    hub: IMessageHub;
    dispatcher: IDispatcher;
    notificationHub: INotificationHub;
    scheduler?: SchedulerLike;
    lifespanMs?: number;
    cap?: number;
  }) {
    super({
      name: opts.name,
      hint: opts.hint,
      hub: opts.hub,
      dispatcher: opts.dispatcher,
    });
    this.#notificationHub = opts.notificationHub;
    this.#scheduler = opts.scheduler ?? asyncScheduler;
    this.#lifespanMs = opts.lifespanMs;
    this.#cap = opts.cap ?? NotificationsVM.DEFAULT_CAP;
  }

  get type(): ViewModelType {
    return ViewModelType.Component;
  }

  get hub(): IMessageHub {
    return this._hub;
  }

  /** Bounded list of currently-rendered notifications. */
  get visible(): readonly NotificationVM[] {
    return this.#visible;
  }

  get cap(): number {
    return this.#cap;
  }

  protected override _onConstruct(): void {
    super._onConstruct();
    this.#pendingSub = this.#notificationHub.pending.subscribe((list) => {
      this.#syncFromPending([...list]);
    });
  }

  protected override _onDestruct(): void {
    this.#pendingSub?.unsubscribe();
    this.#pendingSub = null;
    this.#clearVisible();
    super._onDestruct();
  }

  #syncFromPending(pending: Notification[]): void {
    // Add VMs for new pending notifications, respecting the cap.
    for (const n of pending) {
      if (this.#map.has(n)) continue;
      const vm = new NotificationVM(
        n,
        this.#notificationHub,
        this.#scheduler,
        this.#lifespanMs,
      );
      this.#map.set(n, vm);
      this.#visible.push(vm);
      while (this.#visible.length > this.#cap) {
        const [oldest] = this.#visible.splice(0, 1);
        if (!oldest) continue;
        for (const [k, v] of this.#map) {
          if (v === oldest) {
            this.#map.delete(k);
            break;
          }
        }
        oldest.dispose();
      }
    }
    // Remove VMs whose notifications are no longer pending.
    const stillPending = new Set<Notification>(pending);
    for (const [k, v] of [...this.#map]) {
      if (!stillPending.has(k)) {
        this.#map.delete(k);
        const idx = this.#visible.indexOf(v);
        if (idx >= 0) this.#visible.splice(idx, 1);
        v.dispose();
      }
    }
    this._hub.send(
      PropertyChangedMessage.create(this, this._name, "visible"),
    );
    this._raisePropertyChanged("visible");
  }

  #clearVisible(): void {
    for (const v of this.#visible) v.dispose();
    this.#visible.length = 0;
    this.#map.clear();
  }

  protected override _onDispose(): void {
    this.#pendingSub?.unsubscribe();
    this.#clearVisible();
    super._onDispose();
  }

  static builder(): NotificationsVMBuilder {
    return new NotificationsVMBuilder();
  }
}

export class NotificationsVMBuilder {
  #name: string | null = null;
  #hint = "";
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #notificationHub: INotificationHub | null = null;
  #scheduler: SchedulerLike | null = null;
  #lifespanMs: number | undefined = undefined;
  #cap: number = NotificationsVM.DEFAULT_CAP;

  constructor(from?: NotificationsVMBuilder) {
    if (from) {
      this.#name = from.#name;
      this.#hint = from.#hint;
      this.#hub = from.#hub;
      this.#dispatcher = from.#dispatcher;
      this.#notificationHub = from.#notificationHub;
      this.#scheduler = from.#scheduler;
      this.#lifespanMs = from.#lifespanMs;
      this.#cap = from.#cap;
    }
  }

  name(value: string): NotificationsVMBuilder {
    const b = new NotificationsVMBuilder(this);
    b.#name = value;
    return b;
  }

  hint(value: string): NotificationsVMBuilder {
    const b = new NotificationsVMBuilder(this);
    b.#hint = value;
    return b;
  }

  services(hub: IMessageHub, dispatcher: IDispatcher): NotificationsVMBuilder {
    const b = new NotificationsVMBuilder(this);
    b.#hub = hub;
    b.#dispatcher = dispatcher;
    return b;
  }

  notificationHub(hub: INotificationHub): NotificationsVMBuilder {
    const b = new NotificationsVMBuilder(this);
    b.#notificationHub = hub;
    return b;
  }

  scheduler(scheduler: SchedulerLike): NotificationsVMBuilder {
    const b = new NotificationsVMBuilder(this);
    b.#scheduler = scheduler;
    return b;
  }

  lifespanMs(value: number): NotificationsVMBuilder {
    const b = new NotificationsVMBuilder(this);
    b.#lifespanMs = value;
    return b;
  }

  cap(value: number): NotificationsVMBuilder {
    const b = new NotificationsVMBuilder(this);
    b.#cap = value;
    return b;
  }

  build(): NotificationsVM {
    if (this.#name === null) throw new Error("name is required");
    if (this.#hub === null || this.#dispatcher === null)
      throw new Error("services are required");
    if (this.#notificationHub === null)
      throw new Error("notificationHub is required");
    return new NotificationsVM({
      name: this.#name,
      hint: this.#hint,
      hub: this.#hub,
      dispatcher: this.#dispatcher,
      notificationHub: this.#notificationHub,
      ...(this.#scheduler !== null ? { scheduler: this.#scheduler } : {}),
      ...(this.#lifespanMs !== undefined ? { lifespanMs: this.#lifespanMs } : {}),
      cap: this.#cap,
    });
  }
}

File: examples/csharp/avalonia/NotesShowcase/ViewModels/NotificationsVM.cs
using System.Collections.ObjectModel;
using System.Reactive.Concurrency;
using System.Reactive.Linq;
using VMx.Builders;
using VMx.Components;
using VMx.Messages;
using VMx.Services;
using VMx.Notifications;

namespace NotesShowcase.ViewModels;

/// <summary>
/// Subscribes to an <see cref="INotificationHub"/> and surfaces the most recent
/// notifications as a bounded collection of <see cref="NotificationVM"/>s.
///
/// Cap = 5 (plan §3.a.10). Each <see cref="NotificationVM"/> handles its own
/// auto-dismiss via its <c>Lifespan</c>; this VM observes hub.Pending and
/// removes resolved items.
/// </summary>
public sealed class NotificationsVM : ComponentVMBase
{
    /// <summary>Default maximum number of concurrently rendered notifications.</summary>
    public const int DefaultCap = 5;

    private readonly INotificationHub _notificationHub;
    private readonly IDispatcher _dispatcher;
    private readonly IScheduler _scheduler;
    private readonly TimeSpan? _lifespan;
    private readonly int _cap;
    private readonly ObservableCollection<NotificationVM> _visible = new();
    private readonly Dictionary<Notification, NotificationVM> _map = new();
    private IDisposable? _pendingSub;
    private bool _ownDisposed;

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Component;

    /// <summary>Public hub accessor.</summary>
    public new IMessageHub Hub => base.Hub;

    /// <summary>Bounded list of currently-rendered notifications (newest first).</summary>
    public ObservableCollection<NotificationVM> Visible => _visible;

    /// <summary>Maximum number of concurrently rendered notifications.</summary>
    public int Cap => _cap;

    private NotificationsVM(
        string name,
        string hint,
        IMessageHub hub,
        IDispatcher dispatcher,
        INotificationHub notificationHub,
        IScheduler scheduler,
        TimeSpan? lifespan,
        int cap)
        : base(name, hint, hub, dispatcher, onConstruct: null, onDestruct: null)
    {
        _notificationHub = notificationHub;
        _dispatcher = dispatcher;
        _scheduler = scheduler;
        _lifespan = lifespan;
        _cap = cap;
    }

    /// <inheritdoc/>
    protected override void OnConstruct()
    {
        // live binding: posts arrive from background
        // continuations and auto-dismiss timers (pool threads), and
        // SyncFromPending mutates the ItemsSource-bound collection — INCC
        // off the UI thread crashes Avalonia. Marshal.
        _pendingSub = _notificationHub.Pending
            .ObserveOn(_dispatcher.Foreground)
            .Subscribe(SyncFromPending);
        base.OnConstruct();
    }

    /// <inheritdoc/>
    protected override void OnDestruct()
    {
        _pendingSub?.Dispose();
        _pendingSub = null;
        ClearVisible();
        base.OnDestruct();
    }

    private void SyncFromPending(IReadOnlyList<Notification> pending)
    {
        // Add VMs for new pending notifications, respecting cap.
        foreach (var n in pending)
        {
            if (_map.ContainsKey(n)) continue;
            var vm = new NotificationVM(n, _notificationHub, _scheduler, _lifespan);
            _map[n] = vm;
            _visible.Add(vm);
            // Drop oldest while we're over cap.
            while (_visible.Count > _cap)
            {
                var oldest = _visible[0];
                _visible.RemoveAt(0);
                var keyToRemove = _map.FirstOrDefault(kv => ReferenceEquals(kv.Value, oldest)).Key;
                if (keyToRemove is not null) _map.Remove(keyToRemove);
                oldest.Dispose();
            }
        }
        // Remove VMs whose notifications are no longer pending.
        var stillPending = new HashSet<Notification>(pending, ReferenceEqualityComparer.Instance);
        var toRemove = _map.Keys.Where(k => !stillPending.Contains(k)).ToList();
        foreach (var key in toRemove)
        {
            var vm = _map[key];
            _map.Remove(key);
            _visible.Remove(vm);
            vm.Dispose();
        }
        NotifyPropertyChanged(nameof(Visible));
    }

    private void ClearVisible()
    {
        foreach (var vm in _visible) vm.Dispose();
        _visible.Clear();
        _map.Clear();
    }

    /// <inheritdoc/>
    public override void Dispose()
    {
        if (_ownDisposed) { base.Dispose(); return; }
        _ownDisposed = true;
        _pendingSub?.Dispose();
        ClearVisible();
        base.Dispose();
    }

    /// <summary>Returns a new empty builder.</summary>
    public static NotificationsVMBuilder Builder() => NotificationsVMBuilder.Empty;

    /// <summary>Immutable fluent builder.</summary>
    public sealed class NotificationsVMBuilder
    {
        private readonly string? _name;
        private readonly string _hint;
        private readonly IMessageHub? _hub;
        private readonly IDispatcher? _dispatcher;
        private readonly INotificationHub? _notificationHub;
        private readonly IScheduler? _scheduler;
        private readonly TimeSpan? _lifespan;
        private readonly int _cap;

        internal static readonly NotificationsVMBuilder Empty = new();
        private NotificationsVMBuilder()
        {
            _hint = "";
            _cap = DefaultCap;
        }
        private NotificationsVMBuilder(
            string? name, string hint,
            IMessageHub? hub, IDispatcher? dispatcher,
            INotificationHub? notificationHub, IScheduler? scheduler,
            TimeSpan? lifespan, int cap)
        {
            _name = name; _hint = hint;
            _hub = hub; _dispatcher = dispatcher;
            _notificationHub = notificationHub; _scheduler = scheduler;
            _lifespan = lifespan; _cap = cap;
        }

        /// <summary>Sets the required Name.</summary>
        public NotificationsVMBuilder Name(string name) => new(name, _hint, _hub, _dispatcher, _notificationHub, _scheduler, _lifespan, _cap);
        /// <summary>Sets the optional Hint.</summary>
        public NotificationsVMBuilder Hint(string hint) => new(_name, hint, _hub, _dispatcher, _notificationHub, _scheduler, _lifespan, _cap);
        /// <summary>Sets the required Services.</summary>
        public NotificationsVMBuilder Services(IMessageHub hub, IDispatcher dispatcher) => new(_name, _hint, hub, dispatcher, _notificationHub, _scheduler, _lifespan, _cap);
        /// <summary>Sets the required NotificationHub.</summary>
        public NotificationsVMBuilder NotificationHub(INotificationHub hub) => new(_name, _hint, _hub, _dispatcher, hub, _scheduler, _lifespan, _cap);
        /// <summary>Sets the optional Scheduler for notification timers (default <see cref="DefaultScheduler.Instance"/>).</summary>
        public NotificationsVMBuilder Scheduler(IScheduler scheduler) => new(_name, _hint, _hub, _dispatcher, _notificationHub, scheduler, _lifespan, _cap);
        /// <summary>Overrides the per-notification lifespan (default: 60 s from <see cref="NotificationVM"/>).</summary>
        public NotificationsVMBuilder Lifespan(TimeSpan lifespan) => new(_name, _hint, _hub, _dispatcher, _notificationHub, _scheduler, lifespan, _cap);
        /// <summary>Overrides the cap (default 5).</summary>
        public NotificationsVMBuilder Cap(int cap) => new(_name, _hint, _hub, _dispatcher, _notificationHub, _scheduler, _lifespan, cap);

        /// <summary>Builds the VM after validation.</summary>
        public NotificationsVM Build()
        {
            BuilderValidationException.Require(_name, "Name");
            BuilderValidationException.Require(_hub, "Hub");
            BuilderValidationException.Require(_dispatcher, "Dispatcher");
            BuilderValidationException.Require(_notificationHub, "NotificationHub");
            return new NotificationsVM(
                _name!, _hint, _hub!, _dispatcher!,
                _notificationHub!, _scheduler ?? DefaultScheduler.Instance,
                _lifespan, _cap);
        }
    }
}
